action/protocol/poll/slasher.go

```diff
@@ -440,23 +445,45 @@ func (sh *Slasher) CalculateProbationList(
 	return nextProbationlist, setUnproductiveDelegates(sm, upd)
 }
 
-func (sh *Slasher) calculateUnproductiveDelegatesByEpoch(ctx context.Context, sr protocol.StateReader, rp *rolldpos.Protocol, epochNum uint64) ([]string, error) {
+func (sh *Slasher) calculateUnproductiveDelegates(ctx context.Context, sr protocol.StateReader) ([]string, error) {
 	blkCtx := protocol.MustGetBlockCtx(ctx)
 	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	rp := rolldpos.MustGetProtocol(protocol.MustGetRegistry(ctx))
+	epochNum := rp.GetEpochNum(blkCtx.BlockHeight)
 	delegates, err := sh.GetActiveBlockProducers(ctx, sr, false)
 	if err != nil {
 		return nil, err
 	}
-	numBlks, produce, err := rp.ProductivityByEpoch(epochNum, bcCtx.Tip.Height, sh.productivity)
-	if err != nil {
-		return nil, err
-	}
-	// The current block is not included, so add it
-	numBlks++
-	if _, ok := produce[blkCtx.Producer.String()]; ok {
-		produce[blkCtx.Producer.String()]++
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	var numBlks uint64
+	var produce map[string]uint64
+	if hu.IsPre(config.Goldengate, blkCtx.BlockHeight) {
```

Comment:
can you explain a little bit, what is the new change for this new height?
like calculating productivity in a different way?

Reply:
According to what I discussed with zhi offline, we are thinking about storing latest 720 blockmeta, so that in future it can be used for implementing lightnode as well (not using chainDB but only using state DB). Currently it can replace calculating current productivity logic. So this new change is mainly about writing latest block meta(producer, block time, so on) into stateDB and it will be used in productivity calculation(current epoch) for now, but in future it may have more use cases.